Skipping borrowck because of trivial const

chenyukang opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/149468)*
<!-- homu-ignore:end -->

r? @saethlin

the [assertion](https://github.com/chenyukang/rust/blob/31c38576a4cf1a0864af536f62d11f829db1c7c8/compiler/rustc_mir_transform/src/lib.rs#L424) is added in PR: https://github.com/rust-lang/rust/pull/148040

I think we also need to skip trvial const in `mir_borrowck`.

chenyukang committed:
Skipping borrowck because of trivial const

saethlin commented:
I'm okay with this change, but I would like @lcnr's opinion on whether this sort of whack-a-mole approach is sustainable. (I don't want to have an ever-growing number of `is_trivial_const` checks that eventually upsets the types team)

BoxyUwU commented:
Would definitely be interested in chatting about things here :thinking: @saethlin if you wanna reach out on zulip at some point that'd be cool :)

chenyukang commented:
> > > E.g. is `const FOO: Tait = 1;` a trivial const?
> > 
> > 
> > so this ends up as a trivial const? we never check that the type of the const is not an opaque. This code would start to fail/misbehave with this PR because MIR borrowck doesn't return the defining use anymore, would it not (at least with the new solver)?
> 
> is this an issue with this PR?

I rebased the code with latest upstream.

For `const FOO: Tait = 1;` — the trivial_const function already has a [body.has_opaque_types()](https://github.com/chenyukang/rust/blob/ec79af915ef97e716cf66bede469bd8b07fab169/compiler/rustc_mir_transform/src/trivial_const.rs#L62) check that returns None early, so it's not a trivial const and the borrowck will run.

lcnr commented:
Sorry, marked this as read without seeing your question. I do think this optimization has a very high complexity cost. Having a new kind of body and making sure it's always handled correctly feels brittle to me. I would prefer removing that change :/ I don't remember the perf improvements we get from this and maybe they justify the added complexity :thinking:

> I would expect that the current check isn't sufficient with the new solver here? Because in the new solver we actually normalize `Tait` to its underlying type during writeback, so the body doesn't reference the opaque itself, only when usng the `fn_sig` in borrowck do we actually constrain it.
> 
> Can you test it with `-Znext-solver`?

saethlin commented:
The post-merge perf run is here: https://github.com/rust-lang/rust/pull/148040#issuecomment-3452194818

I think _some_ level of complexity is warranted.

lcnr commented:
yeah :/ i feel conflicted here, which is why I am kind of avoiding this PR :sweat_smile: 


> > For `const FOO: Tait = 1;` — the trivial_const function already has a [body.has_opaque_types()](https://github.com/chenyukang/rust/blob/ec79af915ef97e716cf66bede469bd8b07fab169/compiler/rustc_mir_transform/src/trivial_const.rs#L62) check that returns None early, so it's not a trivial const and the borrowck will run.
> 
> I would expect that the current check isn't sufficient with the new solver here? Because in the new solver we actually normalize `Tait` to its underlying type during writeback, so the body doesn't reference the opaque itself, only when usng the `fn_sig` in borrowck do we actually constrain it.
> 
> Can you test it with `-Znext-solver`?

we can fix this by instead checking whether the constant is the defining scope of any opaque. I guess I am fine with this PR if we add such a test and make sure it works

chenyukang commented:
> yeah :/ i feel conflicted here, which is why I am kind of avoiding this PR 😅
> 
> > > For `const FOO: Tait = 1;` — the trivial_const function already has a [body.has_opaque_types()](https://github.com/chenyukang/rust/blob/ec79af915ef97e716cf66bede469bd8b07fab169/compiler/rustc_mir_transform/src/trivial_const.rs#L62) check that returns None early, so it's not a trivial const and the borrowck will run.
> > 
> > 
> > I would expect that the current check isn't sufficient with the new solver here? Because in the new solver we actually normalize `Tait` to its underlying type during writeback, so the body doesn't reference the opaque itself, only when usng the `fn_sig` in borrowck do we actually constrain it.
> > Can you test it with `-Znext-solver`?
> 
> we can fix this by instead checking whether the constant is the defining scope of any opaque. I guess I am fine with this PR if we add such a test and make sure it works

I changed the trivial-const check so that we do not treat a const as trivial if it is the defining scope of any opaque (`tcx.opaque_types_defined_by(def)`).

I also added a `-Znext-solver` regression test for this scenario. It fails without the new check and passes with it.

lcnr commented on tests/ui/traits/next-solver/opaques/trivial-const-defines-opaque.rs:
```diff
@@ -0,0 +1,13 @@
+//@ compile-flags: -Znext-solver
+//@ check-pass
```
pls also add an old solver revision here

r=me after

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/149468/0469a92a76c327df972cb6c1356934b7a0c6b86d..3c65785e93b8daae8628491e6a3b6c94b9bf97ca)*

chenyukang commented:
@bors r=lcnr

rust-timer commented:
Finished benchmarking commit (f964de49bcb561e5c6c725bb37201e11d852daf0): [comparison URL](https://perf.rust-lang.org/compare.html?start=ffccab6abef68b70c5e23cfbfb1ac7680104040d&end=f964de49bcb561e5c6c725bb37201e11d852daf0&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=ffccab6abef68b70c5e23cfbfb1ac7680104040d&end=f964de49bcb561e5c6c725bb37201e11d852daf0&stat=instructions:u)
Our most reliable metric. Used to determine the overall result above. However, even this metric can be noisy.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.3% | [-0.3%, -0.3%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=ffccab6abef68b70c5e23cfbfb1ac7680104040d&end=f964de49bcb561e5c6c725bb37201e11d852daf0&stat=max-rss)
<details>
<summary>Results (primary 3.3%, secondary -0.3%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 3.3%  | [1.7%, 5.3%]   | 4     |
| Regressions ❌ <br /> (secondary)  | 1.2%  | [0.5%, 2.0%]   | 2     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -3.4% | [-3.4%, -3.4%] | 1     |
| All ❌✅ (primary)                 | 3.3%  | [1.7%, 5.3%]   | 4     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=ffccab6abef68b70c5e23cfbfb1ac7680104040d&end=f964de49bcb561e5c6c725bb37201e11d852daf0&stat=cycles:u)
<details>
<summary>Results (secondary 0.2%)</summary>

A less reliable metric. May be of interest, but not used to determine the overall result above.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 1.0%  | [0.5%, 2.3%]   | 4     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.8% | [-1.1%, -0.5%] | 3     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=ffccab6abef68b70c5e23cfbfb1ac7680104040d&end=f964de49bcb561e5c6c725bb37201e11d852daf0&stat=size:linked_artifact)
This perf run didn't have relevant results for this metric.

**Bootstrap**: 497.124s -> 508.875s (2.36%)
**Artifact size**: 395.05 MiB -> 395.04 MiB (-0.00%)
